packages/programs/program/src/keychain.ts

```diff
@@ -0,0 +1,219 @@
+			| EncryptedSecp256k1Keypair;
+		revoked?: boolean;
+	}) {
+		this.keypair = parameters.keypair;
+		this.owner = parameters.owner;
+		this.revoked = parameters.revoked ?? false;
+	}
+}
+
+export class KeychainProgram extends Program {
+	@field({ type: Documents<EncryptedExtendedKey> })
+	keys: Documents<EncryptedExtendedKey>;
+
+	@field({ type: Identity })
+	identity: Identity;
+
+	@field({ type: EncryptProvide })
+	encryptProvider: EncryptProvide<KeyExchangeOptions>;
+
+	@field({ type: DecryptProvider })
+	decryptProvider: DecryptProvider;
+
+	// TODO: Offer rotate keys functionality. revokes all unrevoked keys and adds a new key which it returns
+	// TODO: Write Key updater that either adds a given key or generates a new one.
+	// TODO: the key updater should always set the revoked flag for keys with a recipient
+
+	// TODO: Write a key getter that get's my latest key. If no valid key is found, run rotate and return newly created
+	// TODO: Write a key getter that get's the latest key's of recipients. If no valid key is found for a recipient, add a new, temporary key for the recipient.
+	// TODO: Write a key getter that get's a specified key, no matter the revoked status
+	async getKey(
+		parameters:
+			| { owner?: Uint8Array; publicKey?: never }
+			| {
+					owner?: never;
+					publicKey?: X25519PublicKey | Ed25519PublicKey | Secp256k1PublicKey;
+			  } = {}
+	) {
+		const { owner = this.identity.publicKey.bytes, publicKey } = parameters;
+		// Create Filter options
+		const queries =
+			// Looking for specific public keys. We don't care about revocation status.
+			publicKey != undefined
+				? [
+						new ByteMatchQuery({
+							key: ["keypair", "publicKey", "publicKey"],
+							value: publicKey.publicKey
+						})
+				  ]
+				: // Looking for latest key from a specific owner.
+				  [
+						new And([
+							// Only non-revoked
+							new BoolQuery({ key: "revoked", value: false }),
+							// Recipient has to be undefined
+							// TODO: Is this correct?
+							new MissingField({ key: "recipient" }),
+							// And from the specified owner
+							new ByteMatchQuery({
+								key: "owner",
+								value: owner
+							})
+						])
+				  ];
+		// TODO: How can we limit this to only fetch the last entry - so a single one?
+		const keys = await this.keys.index.search(
+			new SearchRequest({
+				query: queries,
+				sort: new Sort({
+					key: "timestamp",
+					direction: SortDirection.DESC
+				})
+			})
+		);
+		// Create a new key in case we can't find a match
+		// If owner is not me, create a new, temporary key with me as owner and recipient as the requested owner
+		// If owner is me, create a new permanent key with
+		if (keys.length === 0) {
+			//return compare(this.identity.publicKey.bytes, owner) ?
+		}
+		return keys[0].keypair.decrypt(this.decryptProvider);
+	}
+
+	// TODO: Unsure about encryptProvider here honestly - should this be layer 1 keychain?
+	async open(parameters: {
+		identity: Identity;
```

Comment:
identity already exist if you do

`this.identity` inside open (it is a global thing on 'this')